go-sdk/pkg/bundles/shared/discovery.go

```diff
@@ -0,0 +1,252 @@
+	versions, exists := d.bundles[name]
+	if exists {
+		return versions
+	}
+
+	// We couldn't find the specific named bundle
+
+	// First we see if a "default" was configured
+	versions, exists = d.bundles[viper.GetString("bundles.default_bundle")]
+	if exists {
+		return versions
+	}
+
+	// Else we see if there is exactly one bundle name registered, and if so we return that
+	if len(d.bundles) == 1 {
+		exists = true
+		for key, versions := range d.bundles {
+			// Just pull the first value out
+			d.logger.Debug(
+				"Using sole bundle as fallback bundle",
+				"bundle",
+				name,
+				"fallback_bundle",
+				key,
+			)
+			return versions
+		}
+	}
+	return nil
+}
+
+func (d *Discovery) ClientForBundle(name string, version *string) (*plugin.Client, error) {
+	var key string
+	if version != nil {
+		key = *version
+	}
+
+	versions := d.versionsForNamedBundle(name)
+	if versions == nil {
+		// We couldn't find the specific named bundle
+		return nil, fmt.Errorf(
+			"%wno dag bundle named %q found (and no fallback suitable)",
+			BundleNotFound,
+			name,
+		)
+	}
+
+	cmd, exists := versions[key]
+	if !exists {
+		// We couldn't find the specific version, but lets see if we have just a single version and use that in
+		// its place
+		if key == "" && len(versions) == 1 {
+			exists = true
+			for key, cmd = range versions {
+				// Just pull the first value out
+				d.logger.Info(
+					"Unable to find unversioned bundle as requested, using only version as fallback",
+					"bundle",
+					name,
+					"fallback_version",
+					key,
+				)
+				break
+			}
+		}
+	}
+
+	if !exists {
+		if key == "" {
+			key = "<unversioned>"
+		}
+		return nil, fmt.Errorf("%wno version %q found for dag bundle %q", BundleNotFound, key, name)
+	}
+	return d.makeClient(cmd), nil
+}
+
+func (d *Discovery) DiscoverBundles(ctx context.Context) error {
+	// Find all files in the bundles directory
+	files, err := filepath.Glob(filepath.Join(d.bundlesFolder, "*"))
+	if err != nil {
+		return fmt.Errorf("failed to read bundles directory: %w", err)
+	}
+
+	self, err := os.Executable()
+	if err != nil {
+		self = ""
+	}
+
+	for _, file := range files {
+		if ctx.Err() != nil {
+			// Check if we are done.
+			return ctx.Err()
+		}
+
+		// Check if file is executable
+		if !isExecutable(file) {
+			continue
+		}
+
+		abs, err := filepath.Abs(file)
+		if err != nil {
+			d.logger.Warn("Unable to load resolve file path", "file", file, "err", err)
+			continue
+		}
+		if self != "" && self == abs {
+			d.logger.Warn("Not trying to load ourselves as a plugin", "file", file)
+			continue
+		}
+
+		d.logger.Debug("Found potential bundle", slog.String("path", file))
+
+		// TODO: Use a sync.WaitGroup to parallelize running multiple procs without blowing concurrency up and fork-bombing
+		// the host
+		bundle, err := d.getBundleVersionInfo(file)
+		if err != nil {
+			d.logger.Warn("Unable to load BundleMetadata", "file", file, "err", err)
+			continue
```

Comment:
This would silently skip the bundle if lets say the binary has wrong permissions?

Reply:
This line wouldn't, but L158 

```go
		if !isExecutable(file) {
			continue
		}
```

would. We could add a debug log in there